service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java

```diff
@@ -19,18 +19,157 @@
+    // storage configuration will be found at whatever entity defines it
+    tableProperties.putAll(credentialsMap);
+
+    // Propagate the internal properties to FileIO in case the FileIO implementation needs them
+    Map<String, String> internalProperties =
+        storageInfoEntity.map(PolarisEntity::getInternalPropertiesAsMap).orElse(Map.of());
+    tableProperties.putAll(internalProperties);
+
+    FileIO fileIO = null;
+    fileIO = loadFileIO(impl, tableProperties);
+    return fileIO;
+  }
+
+  private static @Nonnull Optional<PolarisEntity> findStorageInfoFromHierarchy(
+      PolarisResolvedPathWrapper resolvedStorageEntity) {
+    Optional<PolarisEntity> storageInfoEntity =
+        resolvedStorageEntity.getRawFullPath().reversed().stream()
+            .filter(
+                e ->
+                    e.getInternalPropertiesAsMap()
+                        .containsKey(PolarisEntityConstants.getStorageConfigInfoPropertyName()))
+            .findFirst();
+    return storageInfoEntity;
+  }
+
+  private Map<String, String> refreshCredentials(
+      TableIdentifier tableIdentifier,
+      Set<PolarisStorageActions> storageActions,
+      Set<String> tableLocations,
+      PolarisEntity entity) {
+    Boolean skipCredentialSubscopingIndirection =
+        getBooleanContextConfiguration(
+            PolarisConfiguration.SKIP_CREDENTIAL_SUBSCOPING_INDIRECTION.key,
+            PolarisConfiguration.SKIP_CREDENTIAL_SUBSCOPING_INDIRECTION.defaultValue);
+    if (Boolean.TRUE.equals(skipCredentialSubscopingIndirection)) {
+      LOGGER
+          .atInfo()
+          .addKeyValue("tableIdentifier", tableIdentifier)
+          .log("Skipping generation of subscoped creds for table");
+      return Map.of();
+    }
+
+    boolean allowList =
+        storageActions.contains(PolarisStorageActions.LIST)
+            || storageActions.contains(PolarisStorageActions.ALL);
+    Set<String> writeLocations =
+        storageActions.contains(PolarisStorageActions.WRITE)
+                || storageActions.contains(PolarisStorageActions.DELETE)
+                || storageActions.contains(PolarisStorageActions.ALL)
+            ? tableLocations
+            : Set.of();
+    Map<String, String> credentialsMap =
+        entityManager
+            .getCredentialCache()
+            .getOrGenerateSubScopeCreds(
+                getCredentialVendor(),
+                metaStoreSession,
+                entity,
+                allowList,
+                tableLocations,
+                writeLocations);
+    LOGGER
+        .atDebug()
+        .addKeyValue("tableIdentifier", tableIdentifier)
+        .addKeyValue("credentialKeys", credentialsMap.keySet())
+        .log("Loaded scoped credentials for table");
+    if (credentialsMap.isEmpty()) {
+      LOGGER.debug("No credentials found for table");
+    }
+    return credentialsMap;
+  }
+
+  private PolarisCredentialVendor getCredentialVendor() {
+    return metaStoreManager;
```

Comment:
Better narrow the type and change the name of this field - and inline it's usage.